pkg/pillar/cmd/tpmmgr/tpmmgr.go

```diff
@@ -848,6 +850,73 @@ func testEcdhAES() error {
 	return nil
 }
 
+// DecryptWithCipherInfo : Decryption API, for encrypted object information received from controller
+func DecryptWithCipherInfo(cipherBlock types.CipherBlock) (string, error) {
+	// TBD:XXX, for nodes not having tpm chip, the device private key can be used
+	// as a software only solution, which can be wrqpped up
+	// inside DecryptWithEcdhKey
+	if !IsTpmEnabled() {
+		return "", errors.New("Not supported")
+	}
+	if len(cipherBlock.CipherData) == 0 {
+		return "", errors.New("Invalid Cipher Information")
+	}
+	if cipherBlock.KeyExchangeScheme == zconfig.KeyExchangeScheme_KEA_NONE ||
+		cipherBlock.EncryptionScheme == zconfig.EncryptionScheme_SA_NONE {
+		return "", errors.New("No Encryption")
+	}
+	if cipherBlock.KeyExchangeScheme != zconfig.KeyExchangeScheme_KEA_ECDH ||
+		cipherBlock.EncryptionScheme != zconfig.EncryptionScheme_SA_AES_256_CFB {
+		return "", errors.New("Unsupported Encryption protocols")
+	}
+	// currently, its ecdh/aes256
+	cert, err := getControllerCertInfo(cipherBlock)
+	if err != nil {
+		log.Errorf("Could not extract Certificate Information")
+		return "", err
+	}
+	bytes := make([]byte, len(cipherBlock.CipherData))
+	err = DecryptSecretWithEcdhKey(cert.X, cert.Y,
+		cipherBlock.InitialValue, cipherBlock.CipherData, bytes)
+	if err != nil {
+		log.Errorf("Decryption failed with error %v\n", err)
+		return "", err
+	}
+	// TBD:XXX, validate plaintext sha
+	return string(bytes), nil
+}
+
+func getControllerCertInfo(cipherBlock types.CipherBlock) (*ecdsa.PublicKey, error) {
+	var ecdhPubKey *ecdsa.PublicKey
+	if len(cipherBlock.ControllerCert) == 0 || len(cipherBlock.InitialValue) == 0 {
+		return ecdhPubKey, errors.New("Invalid Cipher Information")
+	}
+	// TBD:XXX, validate the ControllerCert Signing
+
+	block := cipherBlock.ControllerCert
+	certs := []*x509.Certificate{}
+	for b, rest := pem.Decode(block); b != nil; b, rest = pem.Decode(rest) {
+		if b.Type == "CERTIFICATE" {
+			c, e := x509.ParseCertificates(b.Bytes)
+			if e != nil {
+				continue
+			}
+			certs = append(certs, c...)
+		}
```

Comment:
And if it is not of type certificate I think you should return an error about the type being unsupported/unknown.